File: 26th/somilee/joystick.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int solution(string name) {
    int ans = 0;
	for(int i = 0; i < name.length(); i++) {
		ans += min(name[i] - 'A', 26 - name[i] + 'A');
	}
	int i = 0;
	int c_move = 1000;
	while(i < name.length()) {
		int n = 0;
		if (name[i] == 'A') {
			while(name[i] == 'A' && i < name.length()) {
				n++;
				i++;
			}
			int a = i - n;
			int b = name.length() - i;
			if (a == 0)
				n = b;
			else n = min(2*(a-1) + 1 + (b-1), 1 + 2*(b-1) + 1 + (a-1));
			c_move = min(n, c_move);
			cout << "c_move = " << c_move << ", i = " << i << '\n';
		}
		i++;
	}
	c_move = min((int)name.length() - 1, c_move);
    return (ans + c_move);
}
int main(void) {
    cout << solution("BBBABBB") << endl;
}

File: 26th/jongmlee/Joystick.java
class Joystick {
	//public static void main(String args[]) {
	//	System.out.println(solution("JOAPOAAAAAAAAQP"));
	//}
	public static int solution(String name) {
		int answer = 0;
		int pre = 0;
		int cnt = Integer.MAX_VALUE;
		for (int i = 0; i < name.length(); i++) {
			char c = name.charAt(i);
			if (c != 'A') {
				answer += Math.min('Z' - c + 1, c - 'A');
				if (i == 0) continue;
				int tmp = pre + name.length() - i;
				cnt = Math.min(cnt, tmp + Math.min(pre, name.length() - i));
				System.out.println("i = " + i + ", tmp = " + tmp + ", pre = " + pre + ", name.length() - i = " + (name.length() - i));
				pre = i;
			}
		}
		cnt = Math.min(pre, cnt);
		return answer + cnt;
	}
}